chemie/events/views.py

```diff
@@ -3,215 +3,390 @@
+            send_event_mail(instance, event)
+            custom_messages = RegistrationMessage.objects.filter(user=instance.user, event=event)
+            for custom_message in custom_messages:
+                messages.add_message(request, messages.INFO, custom_message.message, extra_tags='PS')
+
+        elif status == REGISTRATION_STATUS.WAITING:
+            messages.add_message(request, messages.WARNING, 'Arrangementet er fullt, men du er på venteliste.',
+                                 extra_tags='Venteliste')
+            send_event_mail(instance, event)
+        else:
+            # Denne bør ligge i en try-catch block
+            messages.add_message(request, messages.ERROR, 'En ukjent feil oppstod. Kontakt edb@hc.ntnu.no',
+                                 extra_tags='Ukjent feil')
+
+
+class EditRemoveUserRegistration(SingleObjectMixin, MultiFormsView):
+    model = Event
+    template_name = 'events/deregister_or_edit.html'
+    form_classes = {'deregister': DeRegisterUserForm,
+                    'edit': RegisterUserForm}
+    success_url = 'event:register'
+
+    registration = None
+    object = None
+
+    def get_edit_initial(self):
+        # The boolean fields must be passed to the form along with any instance of current model
+        return {'enable_sleepover': self.object.sleepover,
+                'enable_night_snack': self.object.night_snack,
+                'enable_companion': self.object.companion,
+                'instance': self.registration}
+
+    def dispatch(self, request, *args, **kwargs):
+        # Set event and registration on the whole object. This is run very early
+        self.object = self.get_object()
+        self.registration = EventRegistration.objects.filter(event=self.object, user=self.request.user).first()
+        return super().dispatch(request)
+
+    def get_object(self, queryset=None):
+        obj = self.kwargs.get('object')
+        return obj
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context.update(super(EditRemoveUserRegistration, self).get_context_data(**kwargs))
+
+        # Remove the forms whenever the deadline has passed
+        if not self.object.can_de_register:
+            context.pop('deregister')
+            context.pop('edit')
+
+        # Add queue position
+        registration = self.registration
+        if registration:
+            queue_position = EventRegistration.objects.filter(event=registration.event,
+                                                              created__lt=registration.created,
+                                                              status=REGISTRATION_STATUS.WAITING).count() + 1
+            context.update({'queue_position': queue_position})
+        return context
+
+    def deregister_form_valid(self, form):
+        event = self.object
+
+        # Register the current user and give a slot to the first user in the event queue
```

Comment:
Dette burde vel kanskje være " Deregister the current user .." istedet for "Register the current user .."